packages/opentelemetry-instrumentation-llamaindex/opentelemetry/instrumentation/llamaindex/dispatcher_wrapper.py

```diff
@@ -0,0 +1,144 @@
+            span.set_attribute(f"{SpanAttributes.LLM_PROMPTS}.{idx}.content", message.content)
+
+
+@dont_throw
+def _set_llm_response(event, span) -> None:
+    if should_send_prompts():
+        for idx, message in enumerate(event.messages):
+            span.set_attribute(f"{SpanAttributes.LLM_PROMPTS}.{idx}.role", message.role.value)
+            span.set_attribute(f"{SpanAttributes.LLM_PROMPTS}.{idx}.content", message.content)
+        response = event.response
+        span.set_attribute(
+            f"{SpanAttributes.LLM_COMPLETIONS}.role",
+            response.message.role.value,
+        )
+        span.set_attribute(
+            f"{SpanAttributes.LLM_COMPLETIONS}.content",
+            response.message.content,
+        )
+        if not (raw := response.raw):
+            return
+        span.set_attribute(SpanAttributes.LLM_RESPONSE_MODEL, raw.get("model"))
+        if usage := response.raw.get("usage"):
+            span.set_attribute(SpanAttributes.LLM_USAGE_COMPLETION_TOKENS, usage.completion_tokens)
+            span.set_attribute(SpanAttributes.LLM_USAGE_PROMPT_TOKENS, usage.prompt_tokens)
+            span.set_attribute(SpanAttributes.LLM_USAGE_TOTAL_TOKENS, usage.total_tokens)
+
+
+@dataclass
+class SpanHolder:
+    span: Span
+    token: Any
+    context: context_api.context.Context
+
+
+class OpenLLSpanHandler(BaseSpanHandler[SpanHolder]):
+    _tracer: Tracer = PrivateAttr()
+
+    def __init__(self, tracer: Tracer):
+        super().__init__()
+        self._tracer = tracer
+
+    def new_span(
+        self, id_: str, parent_span_id: Optional[str], **kwargs
+    ) -> Optional[SpanHolder]:
+        """Create a span."""
+        parent = self.open_spans.get(parent_span_id)
+        kind = (
+            TraceloopSpanKindValues.TASK.value
+            if parent
+            else TraceloopSpanKindValues.WORKFLOW.value
+        )
+        # Take the class name from id_ where id_ is e.g.
+        # 'SentenceSplitter.split_text_metadata_aware-a2f2a780-2fa6-4682-a88e-80dc1f1ebe6a'
+        class_name = LLAMA_INDEX_REGEX.match(id_).groups()[0]
+        span_name = f"{class_name}.llama_index.{kind}"
+        span = self._tracer.start_span(
+            span_name,
+            context=parent.context if parent else None,
+            kind=kind,
+        )
+        current_context = set_span_in_context(span, context=parent.context if parent else None)
+        token = context_api.attach(current_context)
+        span.set_attribute(SpanAttributes.TRACELOOP_SPAN_KIND, kind)
+        span.set_attribute(SpanAttributes.TRACELOOP_ENTITY_NAME, span_name)
+        return SpanHolder(span, token, current_context)
+
+    def prepare_to_exit_span(
+        self, id_: str, result: Optional[Any] = None, **kwargs
+    ) -> SpanHolder:
+        """Logic for preparing to drop a span."""
+        span_holder = self.open_spans[id_]
+        span_holder.span.end()
+        context_api.detach(span_holder.token)
```

Comment:
@nirga I added this line as an experiment to see if the spans behave well, but I don't understand it fully. Could you please explain why this is needed?
Without this, the previous span - even if no connection (and `span.end()` already called) - will become the parent when executing `span = self._tracer.start_span(`, which is due to `opentelemetry.sdk.trace.__init__`:
```
parent_span_context = trace_api.get_current_span(
            context
        ).get_span_context()
```
There is a `opentelemetry.context.__init__.detach` call in span.end(), but it's not enough. Is this maybe related to your comment of different spans in opentelemetry and openllmetry? https://github.com/traceloop/openllmetry/issues/1541#issuecomment-2230378905